### DISCUSSION TITLE
How to make Custom transformation on monai in a simple way

### DISCUSSION THREAD

**Discussion by Maryam-Mostafa** [Project-MONAI/MONAI#4809] [Q&A] [answered]
Hi, 
I want to make a custom function for pixelSpacing that takes the a parameter and I want to apply it to the key, and image. I've seen this notebook https://github.com/Project-MONAI/MONAIBootcamp2021/blob/main/day1/1.%20Getting%20Started%20with%20MONAI.ipynb , but couldn't know what to do in my case. 
I don't know if it can be done using lambda ? or if it can be done by making a subclass that inherit maptransform but inside it I'm confused of what should be done here?

this is my funtion :
```
def pixelspacing(img,pxlsp,target_space=[0.4,0.4,0.4]):
   orig_shape= img.shape
   target_shape = [int(np.round(orig_shape[0] *  (target_space[0]/pxlsp[0] ))),
                   int(np.round(orig_shape[1] *  (target_space[1]/pxlsp[1] ))),
                   int(np.round(orig_shape[2] *  (target_space[2]/pxlsp[2] )))]

   resize_factor= np.array(target_shape)/np.array(orig_shape)
   img = ndimage.zoom(img,resize_factor, order=1)
   return img

```
and I want to call it like this in the compose :
` Pixelspacing(("img","seg"),pxlsp)  #pxlsp is the the parameter I want to send to my function`

Thanks in advanced.

**Comment by wyli** [comment, marked as answer]
good question, one way to do it is to use the `adaptor` API in `monai.transforms`
https://github.com/Project-MONAI/MONAI/blob/30d592c33631f15a2cfe27270eb54d1d9087832f/monai/transforms/adaptors.py#L13
```py
import numpy as np
from scipy import ndimage

from monai.transforms import Compose, LoadImageD, adaptor


def PixelSpacing(pxlsp, target_space=[0.4, 0.4, 0.4]):
    def pixelspacing(img, pxlsp=pxlsp, target_space=target_space):
        orig_shape = img.shape
        target_shape = [
            int(np.round(orig_shape[0] * (target_space[0] / pxlsp[0]))),
            int(np.round(orig_shape[1] * (target_space[1] / pxlsp[1]))),
            int(np.round(orig_shape[2] * (target_space[2] / pxlsp[2]))),
        ]

        resize_factor = np.array(target_shape) / np.array(orig_shape)
        img = ndimage.zoom(img, resize_factor, order=1)
        return img

    return pixelspacing


xform = Compose(
    [
        LoadImageD("my_img_key", image_only=True),
        adaptor(PixelSpacing([1.0, 1.0, 1.0]), "output_img_key", {"my_img_key": "img"}),
    ]
)
output = xform({"my_img_key": "./avg152T1_RL_nifti.nii.gz"})
print(output.keys())
print(output["output_img_key"].shape)
```

**Reply by Maryam-Mostafa** [reply]
Thanks you for your help it was really helpful. I tried the proposed code But I got this error when I want to print the output of one image:

`print(train_dataset[0]["img"])`

the error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/monai/transforms/transform.py in apply_transform(transform, data, map_items, unpack_items, log_stats)
     89             return [_apply_transform(transform, item, unpack_items) for item in data]
---> 90         return _apply_transform(transform, data, unpack_items)
     91     except Exception as e:

6 frames
ValueError: operands could not be broadcast together with shapes (3,) (4,) 

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/monai/transforms/transform.py in apply_transform(transform, data, map_items, unpack_items, log_stats)
    112             else:
    113                 _log_stats(data=data)
--> 114         raise RuntimeError(f"applying transform {transform}") from e
    115 
    116 

RuntimeError: applying transform <function adaptor.<locals>._inner at 0x7f24f8b90680>
```

If you have an idea what is the cause of this error or how to fix it I would be really appreciated to you.

**Reply by wyli** [reply]
if you want to use `"img"` then you can replace my demo's `"output_img_key"` and `"my_img_key"` with `"img"`... the demo is trying to distinguish the different keys.

**Reply by Maryam-Mostafa** [reply]
yeah I've already changed this but still getting the same error.
`adaptor(PixelSpacing(pxlsp),"img", {"img": "img"})`
and I also tried this : 
`adaptor(PixelSpacing(pxlsp), ['img'])`
and the same error appears.

**Reply by wyli** [reply]
ok, so I'm not sure how `train_dataset` is created, will need more details to reproduce the error you provided... the error `broadcast together with shapes (3,) (4,)` looks like it's some shape mismatching for example with or without the channel dimension.

**Reply by Maryam-Mostafa** [reply]
I found the error I was applying `AddChannelD` before the `adaptor` and that's why the shape mismatch error occurred. 
I just have one last question, at first I wanted to use the `Spacing  ` method in transforms for making all the images with same pixel size, but I didn't understand how it works. Is there a difference between it and my own implementation could you tell me what is it?
My method is taking the original pixel spacing and the target one then I compute a ratio between them to get a factor that changes the dimensions of the image to reach the target pixel spacing.
Also my target pixel spacing must be in range [0.1,0.5] but the example in the documentation use big values like 5 so I was confusing here.

Thanks for your consideration.

**Reply by wyli** [reply]
looks like the resize factor should be `resize_factor = np.array(orig_shape) / np.array(target_shape)`? still it won't always compute the correct output spatial shape. I'd recommend you use or look into the monai implementation...

**Reply by Maryam-Mostafa** [reply]
Thank you for your generous help and replies 🌹 🙏🏻
